File: src/main/java/runner/Processor.java
package runner;

import technology.idlab.logging.Log;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public abstract class Processor {
    /**
     * Processors which wish to log messages should use the logger provided by
     * the template class. This logger is created with the name of the class
     * which extends the template.
     */
    protected final Log log = Log.Companion.getShared();

    /**
     * The arguments of a processor are stored in a map and can be accessed by
     * name. At the time of writing, the user must manually cast the arguments
     * to the correct type.
     */
    private final Map<String, Object> arguments;

    public Processor(Map<String, Object> arguments) {
        this.arguments = arguments;
    }

    public Processor() {
        this.arguments = new HashMap<>();
    }

    protected <T> T getArgument(String name) {
        Log.Companion.getShared().debug(name);

        Object result = arguments.get(name);

        if (result == null) {
            Log.Companion.getShared().fatal("Argument " + name + " is missing.");
        }

        return (T) result;
    }

    protected <T> Optional<T> getOptionalArgument(String name) {
        Log.Companion.getShared().debug(name + " (optional)");
        return Optional.ofNullable(getArgument(name));
    }

    public void setup() {}

    public void exec() {}
}

File: src/main/kotlin/runner/Processor.kt
package technology.idlab.runner

import java.util.Optional
import technology.idlab.logging.Log

abstract class Processor(
    /**
     * The arguments of a processor are stored in a map and can be accessed by name. At the time of
     * writing, the user must manually cast the arguments to the correct type.
     */
    private val arguments: Map<String, Any>,
) {
  /**
   * Processors which wish to log messages should use the logger provided by the template class.
   * This logger is created with the name of the class which extends the template.
   */
  @JvmField protected val log = Log.shared

  fun <T> getArgument(name: String): T {
    val result = arguments[name] as T ?: log.fatal("Argument $name is missing")
    return result
  }

  fun <T> getOptionalArgument(name: String): Optional<T> {
    val result = arguments[name] as T?
    return Optional.ofNullable(result) as Optional<T>
  }

  abstract fun exec()
}
